### PR TITLE
Add balance strategy

### PR DESCRIPTION
**Pull request by Sengxian**
- Add gate hook
- Add GShard's auxiliary loss
- Add noisy gate balance strategy

### CODE REVIEW

File: fmoe/balance.py [hunk 1]
```diff
@@ -0,0 +1,47 @@
+import torch
+import torch.nn.functional as F
+
+metrics = {
+    "coefficient-variation": lambda c_e: torch.std(c_e) / torch.mean(c_e),
```
Comment (laekov) [inline fmoe/balance.py:5]: need to use either "_" or "-" for all metrices.

File: fmoe/layers.py [hunk 2]
```diff
@@ -212,7 +214,9 @@ def forward(self, inp):
         if self.mp_size > 1:
             inp = Slice.apply(inp, self.mp_rank, self.mp_size, self.mp_group)
 
-        gate_top_k_idx, gate_score = self.gate(inp)
+        gate_top_k_idx, gate_score, gate_state_dict = self.gate(inp)
```
Comment (laekov) [inline fmoe/layers.py:217]: `gate_state_dict` -> `gate_context`

File: fmoe/balance.py [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+    "Lmax_div_Lmin": lambda c_e: (torch.max(c_e) + 1) / (torch.min(c_e) + 1),
+    "Lmax_div_Lmean": lambda c_e: torch.max(c_e) / torch.mean(c_e),
```
Comment (laekov) [inline fmoe/balance.py:7]: div -> over

**Review by laekov — CHANGES_REQUESTED** [review]

**Review by laekov — APPROVED** [review]
LGTM. tql.
